scripts/extract_historical_product_data.pl

```diff
@@ -0,0 +1,335 @@
+--analyze-and-enrich-product-data	run all the analysis and enrichments
+
+TXT
+	;
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Paths qw/%BASE_DIRS/;
+use ProductOpener::Store qw/retrieve store/;
+use ProductOpener::Index qw/:all/;
+use ProductOpener::Display qw/:all/;
+use ProductOpener::Tags qw/:all/;
+use ProductOpener::Products qw/:all/;
+use ProductOpener::Data qw/get_products_collection/;
+use ProductOpener::Ecoscore qw(compute_ecoscore);
+use ProductOpener::LoadData qw/load_data/;
+
+use CGI qw/:cgi :form escapeHTML/;
+use URI::Escape::XS;
+use Storable qw/dclone/;
+use Encode;
+use JSON::PP;
+use Data::DeepAccess qw(deep_get deep_exists deep_set);
+
+use Log::Any::Adapter 'TAP';
+
+use Getopt::Long;
+
+my $analyze_and_enrich_product_data;
+my $codes_file;
+my $field_to_extract;
+my $min_year;
+my $max_year;
+my $recompute_taxonomies;
+my $omit_prefix;
+
+my $query_ref = {};    # filters for mongodb query
+
+GetOptions(
+	# We can extract data for products matching a specific query, or for a list of product codes
+	"query=s%" => $query_ref,    # filters for mongodb query
+	"codes-file=s" => \$codes_file,    # file with product codes to extract data for
+	"field=s" => \$field_to_extract,
+	"omit-prefix" => \$omit_prefix,
+	"recompute-taxonomies" =>
+		\$recompute_taxonomies,    # Recompute tag fields like categories and labels with the current taxonomies
+	"analyze-and-enrich-product-data" => \$analyze_and_enrich_product_data,
+	"min-year=i" => \$min_year,
+	"max-year=i" => \$max_year,
+) or die("Error in command line arguments:\n\n$usage");
+
+my @codes = ();
+
+# Get the list of product codes from a file
+if (defined $codes_file) {
+	open(my $fh, '<', $codes_file) or die("Could not open file '$codes_file' $!");
+	while (my $row = <$fh>) {
+		chomp $row;
+		$row =~ s/\t.*//;    # Assume codes are in the first column; remove the rest
+		push @codes, $row;
+	}
+	close $fh;
+	my $products_count = scalar @codes;
+	print STDERR "$products_count documents retrieved from MongoDB\n";
+
+}
+else {
+	# Use query filters entered using --query categories_tags=en:plant-milks
+
+	use boolean;
+
+	foreach my $field (sort keys %{$query_ref}) {
+
+		my $not = 0;
+
+		if ($query_ref->{$field} =~ /^-/) {
+			$query_ref->{$field} = $';
+			$not = 1;
+		}
+
+		if ($query_ref->{$field} eq 'null') {
+			# $query_ref->{$field} = { '$exists' => false };
+			$query_ref->{$field} = undef;
+		}
+		elsif ($query_ref->{$field} eq 'exists') {
+			$query_ref->{$field} = {'$exists' => true};
+		}
+		elsif ($field =~ /_t$/) {    # created_t, last_modified_t etc.
+			$query_ref->{$field} += 0;
+		}
+		# Multiple values separated by commas
+		elsif ($query_ref->{$field} =~ /,/) {
+			my @tagids = split(/,/, $query_ref->{$field});
+
+			if ($not) {
+				$query_ref->{$field} = {'$nin' => \@tagids};
+			}
+			else {
+				$query_ref->{$field} = {'$in' => \@tagids};
+			}
+		}
+		elsif ($not) {
+			$query_ref->{$field} = {'$ne' => $query_ref->{$field}};
+		}
```

Comment:
nit: would be great to have this kind of things as utility functions for scripts.